### ISSUE TITLE
Use the Cartesian Product to generate IV within a block

### ISSUE THREAD

**Issue by performlabrit** [immersivecognition/unity-experiment-framework#27]
Have you considered using the [Cartesian Product](https://ericlippert.com/2010/06/28/computing-a-cartesian-product-with-linq/) to build blocks?

Here's an example json of an experiment with a practice block and a single main block.

    "blockList": [ "block1","block2"],

    "block1_IV": [ "indepententVarA", "indepententVarB"],
    "block1_indepententVarA": [ 0.5, 0.75, 1.0, 1.25, 1.5 ],
    "block1_indepententVarB": [ 0.4, 0.6, 0.8 ]

    "block2_IV": [ "indepententVarA", "indepententVarB"],
    "block2_indepententVarA": [ 1.5, 1.75, 2.0, 2.25, 2.5 ],
    "block2_indepententVarB": [ 0.4, 0.6, 0.8 ]

If you were to specify a blocklist in your json, it could be used as a prefix to then grab a list of experimental variables per block.  This list could then be used to create the cartesian product of all independent variables (the factorial design).  

I'm not _quite_ sure how to work in repetitions in a manner that provides flexibility to have unequal trial frequencies. Perhaps as another independent variable, e.g. "block1_repetitions": [ 5,5,5,5,5,5,5,5,5,5,5,5,5,5,10], which would have 5 reps of all trials except for the final permutation (block1_indepententVarA = 1.5, block1_indepententVarB = 0.8);

I did take a crack at this myself, but I'm no c# guru, and decided to pass on the challenge once I realized the complexity.

**Comment by jackbrookes** [comment]
Hi,

I think I would normally do nested for loops to achieve this. Detailed here:

https://github.com/immersivecognition/unity-experiment-framework/wiki/Factorial-design

I know its requires a few lines of code compared to a built-in solution, but I would prefer having things be explicitly created rather than

**Comment by performlabrit** [comment]
Yes, it is a suggestion for some automation.  This is such a critical component of experimental design that other labs have developed tools that have focused more specifically on this issue.  Have a look here: https://www.biomotionlab.ca/tux/. The nice thing about their implementation is that it provides a table breakdown of the design the experimenter can visually inspect for issues.

- gD

**Comment by jackbrookes** [comment]
It might not be well documented but UXF has a similar feature, but perhaps not as obvious. after you start the session, you can launch the UXF Session Debugger. I added details here:

**Comment by jackbrookes** [comment]
I think I'll have a go at this just by making a sample Experiment Generator script with an inspector that allows creating factors and their levels. Hopefully users would use this to see its quite easy and more flexible to write your own.
